README
%% /u/sy/beebe/tex/bibclean/bibclean-2.16/README, Tue Jan  1 08:59:06 2013
%% Edit by Nelson H. F. Beebe <beebe@math.utah.edu>
%% /u/sy/beebe/tex/bibclean/2-11/README, Mon Oct  2 10:01:31 1995
%% Edit by Nelson H. F. Beebe <beebe@sunrise>

========================================================================

==========
Jump start
==========

As with most GNUware, you can build, test, and install this program on
most UNIX systems by these simple steps

csh et amici:
	setenv CC ...your favorite C or C++ compiler...
	./configure && make all check install

sh et amici:
	CC=...your favorite C or C++ compiler...
	export CC
	./configure && make all check install

If you don't set the CC environment variable, then gcc (or cc, if gcc
is not available) will be assumed.

If you wish to undo a "make install", just do "make uninstall"; this
will remove any files in system directories put there by "make
install".

See below for further details, and for instructions for non-UNIX systems.
	
This program normally records the username and hostname of the builder;
software distributions may wish to suppress that information by
undefining HOST and USER at in the make command:

	./configure && env HOST= USER= make all check install


============
Introduction
============

This directory contains bibclean, a BibTeX prettyprinter, portability
verifier, and syntax checker.  It can be used to find errors in .bib
files, as well as to standardize their format for readability and
editing convenience.  It can also be used to convert Scribe-format
bibliographies to BibTeX form.

Binary executables for IBM PC DOS, DEC Alpha OpenVMS, DEC VAX VMS, and
Intel x86 Linux may be included in the distribution.

If you do not require either the IBM PC DOS or LINUX, or the DEC VMS
(Alpha and VAX) versions, then you can save about 2.5MB of disk space
by deleting the ibmpc and vms subdirectories.

The default pattern matching in bibclean.c is selected by
HAVE_PATTERNS; with it, no regular-expression library support is
needed.  Should you wish to compile with regular-expression support
instead of the HAVE_PATTERNS code, and your system does not have
compile()/step (HAVE_REGEXP), or re_comp()/re_exec() (HAVE_RECOMP),
you may be able to use the regex-?-??.tar.Z distribution from the Free
Software Foundation, available on prep.ai.mit.edu in /pub/gnu.

In most cases, the HAVE_PATTERNS code is recommended, since it will
give identical results across all machines.  I was prompted to write
it after discovering that there was considerable variety in the
regular expression library codes that resulted in different matching
on different machines, a most unsatisfactory situation.

Please report all problems, suggestions, and comments to the author:

	Nelson H. F. Beebe
	University of Utah
	Department of Mathematics, 110 LCB
	155 S 1400 E RM 233
	Salt Lake City, UT 84112-0090
	USA
	Tel: +1 801 581 5254
	FAX: +1 801 581 4148
	Email: beebe@math.utah.edu, beebe@acm.org, beebe@computer.org (Internet)
	WWW URL: http://www.math.utah.edu/~beebe/


============
Installation
============

Starting with version 2.10.1, bibclean has been adapted to use the GNU
autoconf automatic configuration system for UNIX installations.

GNU autoconf is run at the author's site to produce the configure
script from configure.ac.

The configure script is run at each installer's UNIX site to produce
Makefile from Makefile.in, and config.h from config.hin.  The
configure script is a large (1800+ lines) Bourne shell program that
investigates various aspects of the local C implementation, and
records its conclusions in config.h.  Interestingly, its probes
uncovered a bug in one compiler: lcc 3.4b on Sun Solaris 2.x has an
incorrect definition of toupper() in its ctype.h!

autoconf, at least at the current 2.4 version, is not as C++-aware as
it should be.  The Makefile must carry out minor edits of the
configure script to get it to even work with C++ compilers.  The small
test programs run by configure to determine the existence of assorted
Standard C library functions all lead to incorrect conclusions for
config.h, because they intentionally contain function prototypes with
different argument types.  Since C++ functions are compiled into
external names that encode the function and argument types, along with
the function name, these prototypes produce references to non-existent
functions, causing program linking to fail.  Fortunately, I've been
able to fix this problem too with additional automatic edits, all
carried out by "make configure".  

Should you do a "make maintainer-clean" [NOT recommended, except at the
author's site], the configure script will be deleted, and you will
need recent versions of both GNU m4 and autoconf correctly installed
to reconstruct things, which can be done this way:

	autoconf	# Regenerate unedited configure
	./configure	# Regenerate config.h and Makefile
	rm configure	# delete configure
	make configure	# Regenerate edited configure

For convenience and safety, the distribution includes a subdirectory
named save that contains read-only copies of the files Makefile,
config.h, and configure created by autoconf and "make configure".
This will allow recovery from a lost or damaged configure file.

Suitable hand-crafted config.h files are provided for non-UNIX
systems, and in the unlikely event of a failure of the configure
script on a UNIX system, config.h can be manually produced from a copy
of config.hin with a few minutes' editing work. If you do this,
remember to save a copy of your config.h under a different name,
because running configure will destroy it.  If you have GNU autoconf
installed (the installation is very simple and source code is
available from prep.ai.mit.edu:/pub/gnu/autoconf-x.y.tar.gz), you
might try augmenting config.hin instead, then run autoconf and
configure.

Thus, on UNIX, installation normally consists of just two steps
(assuming a csh-compatible shell):

	setenv CC ...your favorite C or C++ compiler...
	./configure && make all check install

If you like, add OPT='your favorite optimization flags' to the make
command; by default, only -g (debug) is assumed.  If your compiler
won't accept -g with other optimization levels, then set CFLAGS
instead of OPT on the command line; be sure NOT to override any
non-optimizing flags in the CFLAGS set in the Makefile.

The GNU standard installation directories /usr/local/bin for binaries,
and /usr/local/man/man1 for manual pages are assumed.  The prefix
/usr/local can be overridden by providing an alternate definition on
the command line:

	make prefix=/some/other/path install

After installation, you can do
	make distclean
to restore the directories to their distribution state.  You should
also do this between builds for different architectures from the same
source tree; neglecting to do to will almost certainly lead to
failure, because the config.cache file created by configure will lead
to an incorrect config.h for the next build.


============
UNIX Systems
============

The code can be compiled with either C (K&R or ISO/ANSI Standard C) or
C++ compilers.  With some C++ compilers, it may be necessary to supply
additional switches for force the compiler to stay in C++ mode, rather
than reverting to C mode (e.g. on DEC Alpha OSF/1, you must do 
setenv CC "cxx -x cxx").

On UNIX systems, the only changes that you are likely to need in the
Makefile are the settings of CC and CFLAGS, and possibly, DEFINES, and
if you wish to do "make install", the settings of bindir, MANDIR, and
MANEXT.

If you are installing bibclean on a new system, you should definitely
run "make check" before installing it on your system.  For the target
test-bibtex-2, latex is needed.  For test-bibtex-2 and test-scribe-1,
bibtex is needed.  Sample output of "make check" from a UNIX system
is given below.

The code has been tested under more than 55 different C and C++
compilers, and is in regular use to maintain the TeX User Group
bibliography collection stored on ftp.math.utah.edu:/pub/tex/bib, as
well as several other local bibliographies.  These files total more
than 1.08M lines and 62K bibliography entries.  Some of these
bibliographies are mirrored to the Comprehensive TeX Archive Network
(CTAN) hosts.  Do "finger ctan@pip.shsu.edu" to find a CTAN site on
the Internet near you.

bibclean is also used for the BibNet Project, which collects
bibliographies in numerical analysis.  The master collection is
available on ftp.math.utah.edu:/pub/bibnet, and is mirrored from there
to netlib servers at AT&T and Oak Ridge National Laboratory.

If you port bibclean to a new system, please select maximal error and
warning messages in your compiler, to better uncover problems.  If you
find massive numbers of errors complaining about function and argument
type mismatches, it is likely that this can be remedied by suitable
modifications of config.h.  As C implementations move towards
conformance with the December 1989 ISO/ANSI C Language Standard, the C
language is a moving target that must be tracked by config.h, which is
why that file is normally automatically generated on UNIX systems by
the configure script.  With C compilers, you can safely ignore
complaints about implicit declaration of library functions; they are
caused by deficiencies in the vendor-provided header files.

If you have a C++ compiler, please try that as well.  This code has
been successfully compiled under at least 19 C++ compilers, and the
stricter type checking has uncovered problems that slipped past other
compilers.

These programs have been successfully built with C and C++ compilers
and tested on these systems for the 2.10.1 release:

	DECstation 5000		ULTRIX 4.2	cc, gcc, g++, lcc
	DEC Alpha		OSF/1 3.0, 3.2c	cc, c89, cxx, gcc, g++
	HP 9000/375		BSD 4.3		cc, CC
	HP 9000/735		HP-UX 9.0	cc, c89, CC, gcc, g++
	IBM RS/6000		AIX 3.2		cc, c89, xlC, gcc, g++
	Intel 486		Linux 1.3.15	gcc, g++
	MIPS RC6280		RISCos 2.1.1AC	cc
	NeXT 68040		Mach 3.0	cc, cc -ObjC, gcc, g++
	SGI 4D/210		IRIX 4.0.5c	cc, gcc, lcc
	SGI Indigo/2		IRIX 5.3	cc, CC, gcc, g++, lcc
	SGI Power Challenge	IRIX 6.0.1	cc, CC
	Sun SPARCstation	Solaris 2.3,2.4	cc, CC, gcc, g++, lcc
	Sun SPARCstation	SunOS 4.1.3	acc, cc, CC, gcc

Further details are given below.  Where builds have failed, it is
usually because of conflicts between system header files.

The author uses the build-all.sh script for these tests; it tries
builds with every known compiler on the development systems.  If your
UNIX system has other compilers that can be tested, please send their
full path names to the author.


==========
IBM PC DOS
==========

The ibmpc/dos/README file contains details of the builds and tests
of bibclean under 8 IBM PC DOS C and C++ compilers, and instructions
for building and testing bibclean with other compilers.

Since bibclean uses no floating-point arithmetic, and PC DOS has no
shared libraries, I expect that the executables will run on any
version of DOS greater than 4.0.  They may also run on earlier
versions.  At the time of writing, MS-DOS 6.22 is current, and the
bibclean executables work fine on it.


=================
DEC Alpha OpenVMS
=================

The vms/alpha subdirectory contains these files for DEC Alpha OpenVMS:

	bibclean.exe		bibclean executable program
	config.h		hand-coded configuration file

ChangeLog
2019-11-21 10:50:01-0700  Nelson H. F. Beebe  <beebe@math.utah.edu>

	* Prepare, but do not release, bibclean 3.04 with minor changes to
	  address behavior and customization issues raised by a remote
	  user.

	* Makefile.in: Add ${XDEFS} to DEFS value, and add large
	  comment block before XDEFS assignment to document local
	  customizations at some sites.

	* chek.c: Make STD_MAX_TOKEN compile-time settable, and
	  document the reason for the increase in its default value from
	  1000 to 20000 (BibTeX evolution in TeX Live distributions).

	  Add a block enabled by nondefault compile-time definition of
	  DOI_RAW_VALID to accept DOI = "10.xxx" values in BibTeX entries.
	  I do NOT want this to be a default for bibclean, because I feel
	  that BibTeX entries should use the URL form of DOI values, and
	  BibTeX styles that support DOI keywords should strip the URL
	  prefix in the output .bbl file.

	* configure.ac: Update version and date.

	* do.c: In do_initfile(), skip comment lines and empty lines,
	  rather than passing them to do_single_arg() or do_new_pattern().
	  That could be considered a bug fix, but I never noticed
	  unexpected behavior in 29 years of use of bibclean on hundreds
	  of millions of BibTeX entries.

	* token.h: Increase MAX_TOKEN from 32760 to 65525 and
	  revise comment documentation that justifies that change.

2018-03-12 06:44:34-0600  Nelson H. F. Beebe  <beebe@math.utah.edu>

	* Release bibclean 3.03 with no user-visible changes, but with
	  internal redesign to eliminate all memory leaks.  Previously,
	  the user-extendable ISBN and keywords tables contained a mixture
	  of compile-time static strings, and optional user-provided
	  dynamic strings, so their storage was never dynamically
	  recovered.  With the new code in this version, at startup, all
	  static strings in those tables are dynamically reallocated, and
	  then, just before exit, their storage is freed.  Memory leak
	  detectors in clang and gcc (compiler options -fsanitize=address
	  -fsanitize=undefined -fsanitize-recover=all), and Solaris dbx
	  (check -all) now report zero memory leaks when execution
	  terminates.  While those changes might be seen as cosmetic, they
	  will be beneficial if bibclean is again extended in the future,
	  and new memory leaks appear during development and testing.

	* configure.ac: New name for old configure.in to meet recent
	  autoconf requirements.  Update version and PACKAGE_DATE.

	* bibclean.c: Add two function prototypes.  Call
	  keyword_initialize() on startup.  Add new wrapper function
	  finish(status) to call free_keyword_table() and
	  free_ISBN_table() before successful exits, and replace all but
	  two exit(EXIT_xxx) calls by finish(EXIT_xxx).  The exceptions
	  are for fatal errors that are caught early in execution.
	  Failure exits intentionally do not do memory cleanups.

	* bibclean.key: Add rules for ISMN and ORCID-numbers; the latter
	  are present in Web of Science BibTeX output, and record unique
	  check-digited identifiers for authors (see https://orcid.org/
	  and https://en.wikipedia.org/wiki/ORCID).

	* dbx-test.run: New file to automate memory leak checking in
	  Solaris dbx over almost all of the test suite. It uses "check
	  -all" to maximize error checking, but has to use "suppress rui"
	  to suppress read-from-uninitialized checks, because the Solaris
	  C library function, strlen(), raises such errors, perhaps
	  because it acts on words, rather than bytes.

	* do.c: Add new functions free_keyword_table() and
	  keyword_initialize().  Revise add_one_keyword() and
	  add_one_pattern() to free strings before assigning new ones.
	  Change exit() calls to finish() calls to localize memory cleanup
	  immediately before termination.

	* isbn.c: Add new function free_ISBN_table().  Change logic in
	  ISBN_range[] table handling to require (and check for) begin /
	  end / countries triples to be all non--NULL, or all NULL;
	  previously, countries was handled differently.  All assignments
	  to ISBN_range[] table entries are now either NULL, or the result
	  of a Strdup() call, so all non--NULL pointers in that table are
	  allowable in FREE() calls.

	* isbn.h: Add function prototype for free_ISBN_table().

	* match.c: Add condition (p >= &line[0]) in loop to ensure that
	  memory before line[0] is never referenced.

	* option.c: Add prototypes for finish() and free_keyword_table().
	  Change exit() calls to finish() calls to localize memory cleanup
	  immediately before termination.  Update copyright year ranges.

	* isbn.tbl: change all but last entry's NULL pointers to empty
	  strings ("") to match logic changes in ISBN_range[] table
	  handling.

	* isbn-el-to-bibclean-isbn.awk: Change NULL initializations of
	  countries field to empty string ("") to match logic changes in
	  bibclean to remove all memory leaks.

	* test/plain.bst: Copy of standard BibTeX style file so that
	  systems that have a bibtex executable, but lack that style file,
	  or don't have environment variable search paths to find it
	  elsewhere, can find it here during "make check" runs.

2017-06-09 08:20:55-0600  Nelson H. F. Beebe  <beebe@math.utah.edu>

	* Release bibclean 3.02 with minor extensions.

	* chek.c: Revise validate_DOI() to recognize any of a list
	  of DOI prefixes introduced by the DOI agency in 2016.
	  Revise validate_URL() to recognize "https:" protocol,
	  and to issue only one warning message (previously, two
	  similar warnings could be produced).

	* fix.c: Repair an incorrectly-nested multiline comment.

	* configure.in: Update revision date and version number.


match.c
/***********************************************************************

==========
BACKGROUND
==========

This file contains an implementation of limited regular-expression
pattern matching code.  The pattern syntax is simpler, more limited,
and different from normal regular-expression pattern matching syntax.
It is described in more detail below.

The motivation for this new code is that I found considerable
inconsistency in the matching behavior between versions of either
re_comp()/re_exec() or compile()/step() on these systems

	DECstation 3100
	IBM 3090
	IBM PS/2
	IBM RS/6000 AIX 3.2
	NeXT Mach 3.0
	Silicon Graphics IRIX 4.0
	Stardent OS 2.2
	Sun SPARC

That makes use of those regular-expression pattern matching unreliable
across systems.

One possible solution would be to use the GNU re_comp() and re_exec()
from the regexp distribution on prep.ai.mit.edu (as of writing,
pub/gnu/regex-0.11.*).  However, that code is large (5000+ lines), and
its installation uses configuration facilities that only work under
some variants of UNIX, and are completely useless on other operating
systems.

By contrast, the pattern matching code here is quite adequate for
bibclean's needs, and can be expressed in fewer than 140 lines.  In
addition, it provides special handling of TeX control sequences and
braces that would be rather awkward to express in conventional
regular-expression syntax.

If the symbol TEST is defined at compile time, a main program will be
included that can be used for testing patterns supplied from stdin.


==============
PATTERN SYNTAX
==============

The string values to be pattern-matched are tab-free single-line
values delimited by quotation marks.

The patterns are represented by the following special markers:

	a	exactly one letter
	A	one or more letters
	d	exactly one digit
	D	one or more digits
	r	exactly one roman numeral
	R	one or more roman numerals (i.e. a roman number)
	w	exactly one word (one or more letters and digits)
	W	one or more space-separated words, beginning and ending
		with a word
	X	one or more special-separated words, beginning and ending
		with a word
	.	one special character (see SPECIAL_CHARS defined below)
	:	one or more special characters
	<space>	one or more spaces
	\x	exactly one x (x is an character)
	x	exactly the character x (x is anything but aAdDrRwW.:<space>\)

Special characters are a subset of punctuation characters that are
typically used in values.

Note the \<space> represents a single literal space, \\ a single
literal backslash, \a the letter a, \A the letter A, \d the letter d,
\D the letter D, and so on.  Remember to double all backslashes in C
strings: \a must be entered as \\a, and "and" as "\\an\\d".

Each pattern is matched against the entire string and must match
successfully for a YES return from match_pattern().  Consequently,
there is no need for an analogue of ^ and $ in full regular
expressions.  Neither is there provision for matching on arbitrary
sets of characters.  Instead, fixed sets of characters are provided
(conventional regular-expression equivalents are shown in
parentheses):

	digits ([0-9]),
	alphanumerics ([A-Za-z0-9]),
	space ([ \t\f\r\n\v]), and
	special ([][" !#()*+,-./:;?~])

In addition, TeX control sequences of the form
\<one-special-character> or \<letter-sequence> in the string are
ignored in the match, together with any following whitespace.

Braces are also ignored, but not whitespace following them.

Thus "{TR\slash A87}" matches the patterns "AD" and "W", and
"{TR A\slash 87}" matches the patterns "A AD" and "A W".

[29-Jan-1993]
***********************************************************************/

#include <config.h>
#include "xstdbool.h"
#include "xstdlib.h"
#include "xstring.h"
#include "xctype.h"
#include "yesorno.h"
#include "match.h"			/* must come AFTER yesorno.h */

RCSID("$Id: match.c,v 1.10 2017/06/09 14:46:20 beebe Exp beebe $")

#ifndef EXIT_SUCCESS
#define EXIT_SUCCESS 0
#endif

#define RETURN_MATCH_FAILURE(expect) return (match_failure(expect,org_s,s,org_pattern,pattern))

#ifdef MIN
#undef MIN
#endif /* MIN */

#define	MIN(a,b)	((a) < (b) ? (a) : (b))

#define SPECIAL_CHARS	" !#()*+,-./:;?[]~"

#define is_special(c)	IN_SET(SPECIAL_CHARS,(c))

YESorNO	debug_match_failures = NO; 	/* YES: debug value pattern match failures */
static const char	*next_s ARGS((const char *s_));
bool			is_roman ARGS((int c_));
static YESorNO		match_failure ARGS((const char *expect,
					    const char *org_s, const char *s,
					    const char *org_pattern,
					    const char *pattern));
static void		match_warning ARGS((const char *name,
					    const char *org_s, const char *s));
extern void		warning ARGS((const char *msg_));


#if defined(HAVE_STDC)
YESorNO
match_pattern(const char *s, const char *pattern)
#else /* K&R style */
YESorNO
match_pattern(s,pattern)
const char *s;
const char *pattern;
#endif
{
    const char *org_s;
    const char *org_pattern;

    org_s = s;
    org_pattern = pattern;

    s = next_s(s-1);
    for ( ; (*pattern != '\0'); ++pattern)
    {
	switch(*pattern)
	{
	case 'a':			/* single letter */
	    if (!Isalpha((int)*s))
		RETURN_MATCH_FAILURE("single letter");
	    s = next_s(s);
	    break;

	case 'A':			/* one or more letters */
	    if (!Isalpha((int)*s))
		RETURN_MATCH_FAILURE("one or more letters");
	    while (Isalpha((int)*s))
		s = next_s(s);
	    break;

	case 'd':
	    if (!Isdigit((int)*s))	/* single digit */
		RETURN_MATCH_FAILURE("single digit");
	    s = next_s(s);
	    break;

	case 'D':			/* one or more digits */
	    if (!Isdigit((int)*s))
		RETURN_MATCH_FAILURE("one or more digits");
	    while (Isdigit((int)*s))
		s = next_s(s);
	    break;

	case 'r':			/* single roman numeral */
	    if (!is_roman((int)*s))
		RETURN_MATCH_FAILURE("single roman numeral");
	    s = next_s(s);
	    break;

	case 'R':			/* one or more roman numerals */
	    if (!is_roman((int)*s))
		RETURN_MATCH_FAILURE("one or more roman numerals");
	    while (is_roman((int)*s))
		s = next_s(s);
	    break;

	case 'w':			/* one word (letters and digits) */
	    if (!Isalnum((int)*s))
		RETURN_MATCH_FAILURE("one word (letters and digits)");
	    while (Isalnum((int)*s))
		s = next_s(s);
	    break;

	case 'W':			/* one or more space-separated words */
	    if (!Isalnum((int)*s))
		RETURN_MATCH_FAILURE("one or more space-separated words");
	    while (Isalnum((int)*s))		/* parse first word */
		s = next_s(s);
	    for (;;)
	    {
		if (!Isspace((int)*s))
		    break;
		while (Isspace((int)*s))	/* parse separators */
		    s = next_s(s);
		while (Isalnum((int)*s))	/* parse another word */
		    s = next_s(s);
	    }
	    break;

	case 'X':		/* one or more special-separated words */
	    if (!Isalnum((int)*s))
		RETURN_MATCH_FAILURE("one or more special-separated words");
	    while (Isalnum((int)*s))		/* parse first word */
		s = next_s(s);
	    for (;;)
	    {
		if (!is_special(*s))
		    break;
		while (is_special(*s))	/* parse separators */
		    s = next_s(s);
		while (Isalnum((int)*s))	/* parse another word */
		    s = next_s(s);
	    }
	    break;

	case ' ':			/* one or more whitespace characters */
	    if (!Isspace((int)*s))
		RETURN_MATCH_FAILURE("one or more whitespace characters");
	    while (Isspace((int)*s))
		s = next_s(s);
	    break;

	case '.':			/* exactly one special character */
	    if (!is_special(*s))
		RETURN_MATCH_FAILURE("exactly one special character");
	    s = next_s(s);		/* [07-Mar-1999] bug fix: missing before bibclean 2.12 */
	    break;

	case ':':			/* one or more special characters */
	    if (!is_special(*s))
		RETURN_MATCH_FAILURE("one or more special characters");
	    while (is_special(*s))
		s = next_s(s);
	    break;

	case '\\':			/* literal next character */
	    pattern++;
	    /* fall through to exact match test */
	    /*@fallthrough@*/ /*FALLTHROUGH*/

	default:			/* anything else: exact match */
	    if (*pattern != *s)
		RETURN_MATCH_FAILURE("anything else: exact match");
	    s = next_s(s);
	}				/* end switch */
    }					/* end for (; ;) */
    if (*s == '\0')
	return (YES);
    else
	RETURN_MATCH_FAILURE("end of string");
}


#if defined(HAVE_STDC)
static YESorNO
match_failure(const char *expect, const char *org_s, const char *s,
	      const char *org_pattern, const char *pattern)
#else /* K&R style */
static YESorNO
match_failure(expect,org_s,s,org_pattern,pattern)
const char *expect;
const char *org_s;
const char *s;
const char *org_pattern;
const char *pattern;
#endif
{	/* maybe print a warning if -debug-match-failures was specified; always return NO */
    if (debug_match_failures == YES)
    {
#define FORMAT "Pattern does not match %.50s"
	char msg[sizeof(FORMAT) + 50 + 1];

	(void)sprintf(msg,FORMAT,expect);
#undef FORMAT

#ifndef TEST
	warning(msg);
#endif
	match_warning("String ", org_s, s);
	match_warning("Pattern", org_pattern, pattern);
    }
    return (NO);
}


#if defined(HAVE_STDC)
static void
match_warning(const char *name, const char *org_s, const char *s)
#else /* K&R style */
static void
match_warning(name,org_s,s)
const char *name;
const char *org_s;
const char *s;
#endif
{
    char	*msg;

#define DOTS		"..."
#define FORMAT		"%s so far: [%s%.*s], remaining: [%.*s%s]"
#define MAX_FRAG	((size_t)25)	/* longest string fragment allowed */

    msg = (char*)malloc(strlen(FORMAT) + strlen(name) +
			MIN((size_t)(s - org_s),MAX_FRAG) +
			strlen(DOTS) + MIN(strlen(s),MAX_FRAG) +
			strlen(DOTS) + 1);
    if (msg != (char*)NULL)
    {
	const char	*dots1;
	const char	*dots2;

	if ((size_t)(s - org_s) > MAX_FRAG)
	{
	    org_s = s - MAX_FRAG;
	    dots1 = DOTS;
	}
	else
	{
	    dots1 = "";
	}
	if (strlen(s) > MAX_FRAG)
	    dots2 = DOTS;
	else
	    dots2 = "";
	(void)sprintf(msg,FORMAT, name, dots1, (int)(s - org_s), org_s,
		      (int)MIN(strlen(s),MAX_FRAG), s, dots2);
#ifdef TEST
	(void)fprintf(stderr,"%s\n", msg);
#else
	warning(msg);
#endif
	FREE(msg);
    }
#undef FORMAT
}


#if defined(HAVE_STDC)
static const char *
next_s(const char *s)
#else /* K&R style */
static const char *
next_s(s)
const char *s;
#endif
{
    /* find next position in s, ignoring braces and ignoring TeX control
       sequences and any space that follows them */
    for (++s; (*s != '\0'); )
    {
	switch (*s)
	{
	case '\\':			/* TeX control sequence */
	    ++s;			/* look at next character */
	    if (Isalpha((int)*s))		/* \<one-or-more-letters> */
	    {
		while (Isalpha((int)*s))
		    ++s;
	    }
	    else			/* \<non-letter> */
		++s;
	    while (Isspace((int)*s))		/* advance over trailing whitespace */
	        ++s;			/* since TeX does too */
	    break;

	case '{':
	case '}':
	    ++s;
	    break;

	case BIBTEX_HIDDEN_DELIMITER:	/* ignore delimited inline comment */
	    for (++s; (*s != '\0'); ++s)
	    {
		if (*s == BIBTEX_HIDDEN_DELIMITER)
		{
		    ++s;
		    break;
		}
	    }
	    break;

	default:
	    return (s);
	}				/* end switch */
    }					/* end for */
    return (s);
}

#ifdef TEST
#define MAXLINE 256

#define NO_WARNING	(const char *)NULL

MATCH_PATTERN month_patterns[] =
{
    {"aaa",			"oct"},
    {"aaa # \" D\"",		"oct # \" 10\""},
    {"aaa # \" D--D\"",		"oct # \" 20--24\""},
    {"\"D \" # aaa",		"\"10 \" # oct"},
    {"\"D--D \" # aaa",		"\"10--24 \" # oct"},
    {"aaa # \"\" # aaa",	"jul # \"\\emdash \" # aug"},
    {"aaa # \"--\" # aaa",	"jul # \"--\" # aug"},
    {"aaa # \" -- \" # aaa",	"jul # \" -- \" # aug"},
    {"aaa # \"/\" # aaa",	"jul # \"/\" # aug"},
    {"aaa # \" A \" # aaa",	"jul # \" and \" # aug"},
    {(const char*)NULL,		NO_WARNING},
};

MATCH_PATTERN number_patterns[] =
{
    {"\"A AD\"",		"PN LPS5001"},
    {"\"A D(D)\"",		"RJ 34(49)"},
    {"\"A D\"",			"XNSS 288811"},
    {"\"A D\\.D\"",		"Version 3.20"},
    {"\"A-A-D-D\"",		"UMIAC-TR-89-11"},
    {"\"A-A-D\"",		"CS-TR-2189"},
    {"\"A-A-D\\.D\"",		"CS-TR-21.7"},
    {"\"A-AD-D\"",		"TN-K\\slash 27-70"},
    {"\"A-D D\"",		"PB-251 845"},
    {"\"A-D-D\"",		"ANL-30-74"},
    {"\"A-D\"",			"TR-2189"},
    {"\"AD-D-D\"",		"GG24-3611-00"},
    {"\"AD-D\"",		"SP43-29"},
    {"\"AD\"",			"LPS0064"},
    {"\"A\\#D-D\"",		"TR\\#89-24 ????"},
    {"\"D  D\"",		"23 \\& 24"},
    {"\"D \\an\\d D\"",		"11 and 12"},
    {"\"D+D\"",			"3+4"},
    {"\"D-D\"",			"23-27"},
    {"\"D/D\"",			"23/27"},
    {"\"DA\"",			"23A"},
    {"\"D\"",			"23"},
    {"\"D\\.D\"",		"3.4"},
    {"\"W-W W\"",		"AERE-R 12329"},
    {"\"W-W-WW-W\"",		"OSU-CISRC-4\\slash 87-TR9"},
    {"\"W\"",			"Computer Science Report 100"},
    {"\"X\"",			"TR/AB/3-43.7-3/AB"},
    {(const char*)NULL,		NO_WARNING},
};

MATCH_PATTERN pages_patterns[] =
{
    {"\"D\"",			"23"},
    {"\"aD\"",			"L23"},
    {"\"D--D\"",		"23--27"},
    {"\"aD--aD\"",		"L23--L27"},
    {"\"D, D\"",		"23, 27"},
    {"\"aD, aD\"",		"L23, L27"},
    {"\"D, D, D\"",		"23, 27, 45"},
    {"\"aD, aD, aD\"",		"L23, L27, L45"},
    {"\"D, D, D, D\"",		"23, 27, 45, 98"},
    {"\"aD, aD, aD, aD\"",	"L23, L27, L45, L98"},
    {"\"R + D\"",		"viii + 445"},
    {"\"R + D, w D w\"",	"viii + 445, with 30 illustrations"},
    {"\"D, w D w\"",		"239, with 27 illustrations"},
    {"\"D--D, D--D\"",		"23--27, 29--32"},
    {"\"D--D, D--D, D--D\"",	"23--27, 29--32, 35--37"},
    {"\"aD--aD, aD--aD\"",	"L23--L27, L29--L32"},
    {"\"aD--aD, aD--aD, aD--aD\"", "L23--L27, L29--L32, L35--L37"},
    {(const char*)NULL,		NO_WARNING},
};

MATCH_PATTERN volume_patterns[] =
{
    {"\"D\"",			"27"},
    {"\"DA\"",			"27A"},
    {"\"D/D\"",			"27/3"},
    {"\"DA D\"",		"27A 3"},
    {"\"w-D\"",			"SMC-13"},
    {"\"A\"",			"VIII"},
    {"\"D\\.D\"",		"1.2"},
    {"\"D \\an\\d D\"",		"11 and 12"},
    {"\"W\"",			"Special issue A"},
    {(const char*)NULL,		NO_WARNING},
};

MATCH_PATTERN year_patterns[] =
{
    {"\"DDDD\"",		NO_WARNING},
    {"\"DDDD,WDDDD\"",		NO_WARNING},
    {"\"DDDD, DDDD, DDDD\"",	NO_WARNING},
    {"\"18dd, 18dd, 18dd\"",	"1889, 1890, 1891"},
    {"\"18dd, 18dd\"",		"1889, 1890"},
    {"\"18dd--d\"",		"1891--2"},
    {"\"18dd\"",		"1892"},
    {"\"18dda18dd\"",		"{\\noopsort{1885a}}1885"},
    {"\"19dd (19dd)\"",		"1989 (1990)" },
    {"\"19dd, 19dd, 19dd\"",	"1989, 1990, 1991"},
    {"\"19dd, 19dd\"",		"1989, 1990"},
    {"\"19dd--d\"",		"1991--2"},
    {"\"19dd\" # \"--\"",	"\"1989\" # \"\\unskip--\""},
    {"\"19dd\"",		"1992"},
    {"\"19dda19dd\"",		"{\\noopsort{1985a}}1985"},
    {"\"200d--d\"",		"2001--2"},
    {"\"200d\"",		"2009"},
    {(const char*)NULL,		NO_WARNING},
};

static long	line_number;

int		main ARGS((int argc,char* argv[]));
static void	process ARGS((const char *line_, MATCH_PATTERN patterns_[]));

#if defined(HAVE_STDC)
int
main(int argc, char* argv[])
#else /* K&R style */
int
main(argc,argv)
int argc;
char* argv[];
#endif
{
    char line[MAXLINE];

    /* Input lines should look like
       		key = "value",
       where key is one of: month, number, pages, volume, or
       year. Lines with any other key values are ignored. */

    line_number = 0L;
    while (fgets(line,MAXLINE,stdin) != (char*)NULL)
    {
	char *p;

	line_number++;

	p = strchr(line,'\0');
	if (p != (char *)NULL)
	{
	    for (--p; (p >= &line[0]) && (Isspace((int)*p) || (*p == ',')); --p)
		*p = '\0';	/* remove trailing whitespace and commas */
	    for (p = line; Isspace((int)*p); ++p)
       		continue;
	    if (strncmp(p,"month",4) == 0)
		process(line,month_patterns);
	    else if (strncmp(p,"number",6) == 0)
		process(line,number_patterns);
	    else if (strncmp(p,"pages",4) == 0)
		process(line,pages_patterns);
	    else if (strncmp(p,"volume",6) == 0)
		process(line,volume_patterns);
	    else if (strncmp(p,"year",4) == 0)
		process(line,year_patterns);
	    else
       		printf("%%%% %ld [%-24s]: %s\n", line_number, line, "ignored");
	}
    }

    exit (EXIT_SUCCESS);
    return (EXIT_SUCCESS);
}


#if defined(HAVE_STDC)
static void
process(const char *line, MATCH_PATTERN patterns[])
#else /* K&R style */
static void
process(line,patterns)
const char *line;
MATCH_PATTERN patterns[];
#endif
{
    int k;
    const char *p;

    p = strchr(line,'\"');

    if (p != (char *)NULL)
	line = p;

    for (k = 0; patterns[k].pattern != (const char*)NULL; ++k)
    {
	if (match_pattern(line,patterns[k].pattern) == YES)
	{
	    if (patterns[k].message != NO_WARNING)
		printf("%%%% %ld [%-24s]: matches %s\n", line_number, line,
		       patterns[k].message);
	    return;
	}
    }

    printf("?? %ld [%-24s]: illegal value\n", line_number, line);
}
#endif /* TEST */
